src/control/dmg/utils.go

```diff
@@ -80,6 +80,7 @@ func annotateState(state *pb.ResponseState) {
 // unpackClientMap takes a map of addresses to result type and prints either
 // decoded struct or provided error.
 func unpackClientMap(i interface{}) string {
+	var answer interface{}
```

Comment:
I realize that it would be a much larger and invasive change, but why can't this be done by implementing the fmt.Stringer interface on the various map types? I will fully admit that I am still a ways away from being able to claim that I understand the whole code base, but the use of interface{} is a strong code smell. Yes, it works, but there is almost a better way to do it.

Reply:
I'm not well versed in go so this might be a dumb question. But what would be the underlying problem if interface{} is continued to be used in this way?

Reply:
do you want a re-factor on this patch or follow-up?

Reply:
It's not a dumb question at all! I feel bad for picking on Tom's code, but I felt that it would be better to raise this sooner rather than later. It's not wrong per se to use interface{} -- it's kind of like void * in C. There are times when it's the right way to do things, but it should ideally be used sparingly because it doesn't say anything. Rob Pike has a proverb for this, actually -- "The empty interface says nothing. ... If you're going to use it, you might as well be writing Python." The issue is that by accepting an empty interface as a function parameter, you're not imposing any guarantees on the caller, so the caller can give your function things you didn't expect and the compiler won't help you out. In this particular instance I understand why the code is the way it is -- ultimately these maps are stringified using the yaml library and this removes some of the boilerplate.

A different approach might be to get rid of unpackClientMap() altogether and instead delegate responsibility for stringification to the types themselves. Something along the lines of this:

```go
// FeatureMap is an alias for mgmt features supported by gRPC server.
type FeatureMap map[string]string

func (fm FeatureMap) String() string {
	var sb strings.Builder

	for k, v := range fm {
		sb.WriteString(fmt.Sprintf("%s: %s\n", k, v))
	}

	return sb.String()
}

// FeatureResult contains results and error of a request
type FeatureResult struct {
	Fm  FeatureMap
	Err error
}

func (fr FeatureResult) String() string {
	if fr.Err != nil {
		return fr.Err.Error()
	}
	return fr.Fm.String()
}

// ClientFeatureMap is an alias for management features supported on server
// connected to given client.
type ClientFeatureMap map[string]FeatureResult

func (cfm ClientFeatureMap) String() string {
	var sb strings.Builder

	for server, featureMap := range cfm {
		sb.WriteString(fmt.Sprintf("%s:\n", server))
		for _, feature := range strings.Split(featureMap.String(), "\n") {
			sb.WriteString(fmt.Sprintf("    %s\n", feature))
		}
	}

	return sb.String()
}
```
There would be a tradeoff for writing more code countered by having more control over how each type is turned into a string representation as well as keeping the stringification code near the type definition and other type methods. It also has a potential benefit of removing the dependency on yaml, at least in this area of the code.

For the record: I'm well aware of the fact that I'm wading into an established codebase with lots of history and good reasons for why things are the way they are now, so I'm mostly just trying to make suggestions while I still have an outsider's perspective. I don't think it's worth blocking this PR over it.

Reply:
Thanks for the explanation @mjmac, I see now why it's a code smell.

Reply:
very good reflections, you are right that this could be optimised and I completely agree with the proverb. using interface as a parameter is a cheat and implemented here because really I wasn't very concerned about making it robust because I considered it throwaway and just needed to print the output rather than store anything useful.

given this PR doesn't actually implement the approach, only adjusts it, it could be said that adding Stringer implementation is work for another PR, having said that I am also happy to refactor. it's your call :-) thanks for the comments and keep them coming, we can all learn.

Reply:
Absolutely -- I don't really think I'm in a position to be gatekeeping anything just yet; just wanted to point this out as it went by. I think addressing the stringer issue in a different PR is reasonable.

Reply:
would you mind copying your comment above into a ticket and adding to the backlog? can assign it to me.

Reply:
https://jira.hpdd.intel.com/browse/DAOS-2657 -- I think I added the correct labels/status...